component/loki/source/docker/internal/dockertarget/target.go

```diff
@@ -200,24 +203,8 @@ func (t *Target) process(r io.Reader, logStream string) {
 			continue
 		}
 
-		// Add all labels from the config, relabel and filter them.
-		lb := labels.NewBuilder(nil)
-		for k, v := range t.labels {
-			lb.Set(string(k), string(v))
-		}
-		lb.Set(dockerLabelLogStream, logStream)
-		processed, _ := relabel.Process(lb.Labels(), t.relabelConfig...)
-
-		filtered := make(model.LabelSet)
-		for _, lbl := range processed {
-			if strings.HasPrefix(lbl.Name, "__") {
-				continue
-			}
-			filtered[model.LabelName(lbl.Name)] = model.LabelValue(lbl.Value)
-		}
-
 		t.handler.Chan() <- loki.Entry{
-			Labels: filtered,
+			Labels: logStreamLset,
```

Comment:
The only concern I'd have here is that may be we should not give away a shared, mutable copy of this. Could use `maps.Copy()`.

Reply:
To be honest, we've had a similar [issue](https://github.com/grafana/agent/pull/3861) in the past, but given the overall [promise](https://github.com/grafana/agent/blob/4b8807a517071556c7fd520d7ddd792fbcd311f7/component/component.go#L36-L45) to treat arguments as immutable we decided to only copy/clone on the downstream components that could change the label set (today, only loki.process and loki.relabel), so I don't think it's necessary to do it here as well. WDYT?

Reply:
Also, the idea is that we could further reduce the copying-related resource usage with a copy-on-write model eg. only for actual relabeling rules or processing stages that _do_ change the labels, instead of doing it preemptively on every entry passing through.
https://github.com/grafana/agent/blob/8bcc595ecea6b41a83193065a91be17395bb5b91/component/loki/process/process.go#L139-L144

Reply:
Fair enough - I'm happy with this approach if that's our deliberate choice :) 

One day maybe we could have an immutable wrapper that would not copy? 🤔 not sure if that would be more efficient than copying in our case and how would the implementation look in Go.